dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/groups/impl/explicit/ExplicitGroupProvider.java

```diff
@@ -99,33 +109,71 @@ public ExplicitGroup makeGroup() {
     RoleAssignee findRoleAssignee(String roleAssigneeIdtf) {
         return roleAssigneeSvc.getRoleAssignee(roleAssigneeIdtf);
     }
-
+    
+    
+    public boolean contains(DataverseRequest req, ExplicitGroup explicitGroup) {
+        return containsDirectly(req, explicitGroup) || containsIndirectly(req, explicitGroup);
+    }
+    
+    
     /**
-     * Sets the provider of the passed explicit group to {@code this}.
-     *
-     * @param eg the collection
-     * @return the passed group, updated.
+     * @param req
+     * @return {@code true} iff the request is contained in the group or in an included non-explicit group.
      */
-    ExplicitGroup updateProvider(ExplicitGroup eg) {
-        if (eg == null) {
-            return null;
+    protected boolean containsDirectly(DataverseRequest req, ExplicitGroup explicitGroup) {
+        User ra = req.getUser();
+        if (ra instanceof AuthenticatedUser) {
+            AuthenticatedUser au = (AuthenticatedUser) ra;
+            if (explicitGroup.getContainedAuthenticatedUsers().contains(au)) {
+                return true;
+            }
+        }
+
+        if (explicitGroup.getContainedRoleAssignees().contains(ra.getIdentifier())) {
+            return true;
+        }
+
+        for (String craIdtf : explicitGroup.getContainedRoleAssignees()) {
+            // Need to retrieve the actual role assingee, and let it's logic decide.
+            RoleAssignee cra = findRoleAssignee(craIdtf);
+            if (cra != null) {
+                if (cra instanceof Group) {
+                    Group cgrp = (Group) cra;
+                    if (otherGroupProviderContains(req, cgrp, cgrp.getClass())) {
+                        return true;
+                    }
+                } // if cra is a user, we would have returned after the .contains() test.
+            }
         }
-        eg.setProvider(this);
-        return eg;
+        // If we get here, the request is not in this group.
+        return false;
     }
 
     /**
-     * Sets the provider of the explicit groups to {@code this}.
-     *
-     * @param <T> Collection's type
-     * @param egs the collection
-     * @return the collection, with all the groups updated.
+     * @param req
+     * @return {@code true} iff the request if contained in an explicit group that's a member of this group.
      */
-    <T extends Collection<ExplicitGroup>> T updateProvider(T egs) {
-        for (ExplicitGroup eg : egs) {
-            updateProvider(eg);
+    private boolean containsIndirectly(DataverseRequest req, ExplicitGroup explicitGroup) {
+        for (ExplicitGroup ceg : explicitGroup.getContainedExplicitGroups()) {
+            if (this.contains(req, ceg)) {
+                return true;
+            }
+        }
+        return false;
+    }
+    
+    @SuppressWarnings("unchecked")
```

Comment:
Can you add a comment explaining why warning is suppressed?

Reply:
ok